### PR TITLE
Abstractioned API

### PR DESCRIPTION
**Pull request by ygins**
- Abstracted away the DataStore into "base" package
- Store items as strings with serializers/deserializers
- key publisher/retriever work on top of datastores
- Added --sshdir as cli to publisher

### CODE REVIEW

File: switcheroo/base/data_store/s3.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+from switcheroo.base.data_store import DataStore
+
+T = TypeVar("T")
+
+
+class S3DataStore(DataStore):
+    def __init__(self, _bucket_name: str):
+        super().__init__()
+        self._bucket_name = _bucket_name
+        self._s3_client = boto3.client("s3")  # type: ignore
```
Comment (DbarYU) [inline switcheroo/base/data_store/s3.py:18]: region of the S3 bucket?

**Review by yydiamen — APPROVED** [review]

File: switcheroo/paths.py [hunk 2]
```diff
@@ -19,6 +19,14 @@ def local_public_key_loc(host: str, user: str, home_dir: Path | None) -> Path:
     return local_key_dir(host, user, home_dir=home_dir) / KeyGen.PUBLIC_KEY_NAME
 
 
+def local_relative_public_key_loc(host: str, user: str) -> Path:
```
[the remark this answers is not in the document]
Comment (ygins) [reply switcheroo/paths.py:22]: hmmm not sure. What would that util method be? _host_user(host, user)? At that point we're basically back to where we started - code repetition with the same method. I think this is more explicit

File: switcheroo/ssh/objects/key.py
Comment (DbarYU) [inline switcheroo/ssh/objects/key.py:1]: Shouldnt Key and KeyMetaData be coupled? maybe Key takes KeyMetaData object as a constructor

File: switcheroo/ssh/data_org/publisher/__init__.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    ) -> tuple[Key, KeyMetadata]:
+        # Lazy evaluation of default values
+        if key is None:
+            key = Key()
+        if metadata is None:
+            metadata = KeyMetadata.now_by_executing_user()
+        self.publish_public_key(key.public_key, host, user)
+        self.publish_private_key(key.private_key, host, user)
+        self.publish_key_metadata(metadata, host, user)
+        return (key, metadata)
```
Comment (DbarYU) [inline switcheroo/ssh/data_org/publisher/__init__.py:36]: Should not return just key with metadata encapsulated in?

File: switcheroo/ssh/objects/key.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+from typing import IO, ClassVar
+from datetime import datetime
+from switcheroo.base import Serializer
+from switcheroo.custom_keygen import KeyGen
+
+
+class Key:  # pylint: disable=too-few-public-methods
+    class PrivateComponent:
+        def __init__(self, byte_data: bytes):
+            self.byte_data = byte_data
```
Comment (DbarYU) [inline switcheroo/ssh/objects/key.py:13]: , shouldn't keys components be encrypted?

File: switcheroo/ssh/objects/key.py
Comment (ygins) [reply switcheroo/ssh/objects/key.py:1]: You make a good point I think

File: switcheroo/ssh/data_stores/__init__.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+def sshify(data_store: T) -> T:
+    data_store.register_serializer(Key.PrivateComponent, PrivateKeySerializer())
+    data_store.register_serializer(Key.PublicComponent, PublicKeySerializer())
+    data_store.register_serializer(KeyMetadata, KeyMetadataSerializer())
+    if isinstance(data_store, FileDataStore):
+        data_store.register_file_permissions(
+            Key.PrivateComponent, FileDataStore.FilePermissions(0o600)
+        )
+    return data_store
+
```
Comment (DbarYU) [inline switcheroo/ssh/data_stores/__init__.py:24]: I think this breaks a bit of the generic implementation of the function.

Comment (ygins) [reply switcheroo/ssh/data_stores/__init__.py:24]: I dont think so because it is ssh specific

**Pull request merged**